#1v1kddn [Postal Address Validation]

ivanganatra opened the pull request:
-  Address Validation Service. [validate#PostalAddress]

- Address Map service for creating request Maps. [get#AddressMap]

***Related Ticket Link:[https://app.clickup.com/t/1v1kddn](url)**

ivanganatra committed:
Set up initial structure of mantle-fedex

ivanganatra committed:
Update FedexServices.xml

Removed Valid Postal Address Service from initial setup.

ivanganatra committed:
Created maps using API Token and stored response

ivanganatra committed:
Changed name of Api Token

ivanganatra committed:
Added get and store services for authorization.

ivanganatra committed:
Update FedexServices.xml

Added Sample Address in Service Description

ivanganatra committed:
Added Description of Validate Postal Address Service

ivanganatra committed:
Merge branch 'AddressValidation' of github.com:ivanganatra/mantle-fedex into AddressValidation

ivanganatra committed:
Changed folder name of entities to entitiy

ivanganatra committed:
Default Services and Entities for Authentication Added

ivanganatra committed:
Moved default services above the main services

ivanganatra committed:
Implemented OrderShippingRateAPI, added accountNumber in token details

ivanganatra committed:
Added Initial Authentication SetupData for token creation

ivanganatra committed:
Merged main and RateAndTransitTime Branch

ivanganatra committed:
Merged Authentication Details from Main Branch

ivanganatra committed:
Demo Postal Data Added, Validation service now verifies contactMechId's Address

ivanganatra committed:
Implemented get#AddressMap Service, converts Address to RequestMap

ivanganatra committed:
Removed dependency of mantle

ivanganatra committed:
GeoData Corrected, party table added, token stored in Shipping Entity"

ivanganatra committed:
Connection Service Added

ivanganatra committed:
Comments and Alignment Corrected

ivanganatra committed:
Created branch for Address Validation Service

ivanganatra committed:
Removed Unwanted Code and Log Messages

ivanganatra committed:
Updated Postal Address, and returned new contactMechId

ivanganatra committed:
Log messages corrected

ivanganatra commented:
@dixitdeepak Please review this PR.

ivanganatra committed:
Added facility Details, corrected indentation and added comments

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,159 @@
+<services xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://moqui.org/xsd/service-definition-2.1.xsd">
+
+    <service verb="validate" noun="PostalAddress">
+        <in-parameters>
```
Please check the use of interface, we should implement the existing validate ShippingPostalAddress interface and add our custom validate service rather than adding the parameters separately.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,159 @@
+
+            <!-- Setting Connection With Fedex -->
+            <service-call name="mantle.fedex.FedexServices.set#Connection"
+                          in-map="[method:'POST', connectUrl:'https://apis-sandbox.fedex.com/address/v1/addresses/resolve',
```
You should not hard code the connection url, please check to store the fedex connection details.

ivanganatra committed:
Corrected alignment, set required attribute to true

ivanganatra committed:
Deleted MoquiConf.xml file

ivanganatra committed:
Renamed DemoData file

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,149 @@
+
+            <!-- Converting Address Data to JSON Request Map -->
+            <service-call name="mantle.fedex.FedexServices.get#AddressMap" in-map='[contactMechId:contactMechId, partyId:partyId]'
+                    out-map='addressMap'/>
```
Please be consistent with the formatting, single quotes work but use double quotes to align with OOTB code patterns.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,149 @@
+        <actions>
+            <entity-find-one entity-name="mantle.party.contact.PartyContactMech" value-field="partyContactMech">
+                    <field-map field-name="contactMechId" from="contactMechId"/>
+                    <field-map field-name="contactMechId" from="contactMechId"/>
```
Why is field-map mentioned twice? Please check. Will the entity-find-one work if we dont mention field-map here at all?

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,149 @@
+            </entity-find-one>
+
+            <!-- Creating AddressMap -->
+            <set field="adrs" from="streetLines:[postalAddress.address1],
```
This type of naming is not preferred, use better name in place of adrs.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,149 @@
+                                    countryCode:geoCountryData.geoCodeAlpha2,
+                                    postalCode:postalAddress.postalCode" type="NewMap"/>
+            <set field="addressMap" from="[addressesToValidate:[[address:adrs]]]" type="NewMap"/>
+            <set field="validContactMechId" value="true"/>
```
Are we using this value somewhere? What is the purpose of this?

ivanganatra committed:
Now address not validated if contactMechId is expired, removed update#PostalAddress service(service now updates only trustLevelEnumId) and removed partyId parameter from get#AddressMap

ivanganatra committed:
Updated trustLevelEnumId with the attributes returned for verifying that address is validated, renamed geoStateData to geoStateProvinceData as it works for both Stateand Province

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,116 @@
+
+            <log message="Fedex validate address ${contactMechId} trust ${trustLevelEnumId}}"/>
+            <!-- Updating trustLevelEnumId value -->
+            <entity-find-one entity-name="mantle.party.contact.ContactMech" value-field="contactMech"/>
```
I think the update is done twice by mistake here. Please check.

ivanganatra committed:
Corrected indentation

ivanganatra committed:
Removed duplicate storing of trust level enum id

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,112 @@
+            </if>
+
+            <!-- Verifying Correctness of Address -->
+            <set field="responseMap" from="response.responseMap"/>
```
If addressMap.isAddressComplete is false, this is potential NPE as the control flow reaches this code. Please check for proper handling.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,112 @@
+
+            <!-- Setting Connection With Fedex -->
+            <if condition="addressMap.isAddressComplete=='true'">
+            <service-call name="mantle.fedex.FedexServices.send#FedexRequest" out-map="response"
```
Check code indentation. You must follow the best practices for code formatting.

ivanganatra committed:
Corrected alignment
Removed isAddressCompleteParameter
Correct statusCode to statusCodes and CmtValid to CmtlValid
Made get#AddressMap generic, it returns contactMap and addressMap

ivanganatra committed:
Corrected errors by testing in Address Service
Added error messages for AddressValidation service

ivanganatra committed:
Merged address map and contact map into one map, added all alerts, corrected name and organization details

ivanganatra committed:
Removed tokenValue from input while calling send#FedexRequest, now token will be searched in send#FedexRequest only

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,145 @@
+            <!-- Converting Address Data to JSON Request Map -->
+            <service-call name="mantle.fedex.FedexServices.get#AddressMap" in-map="[postalContactMechId:contactMechId]"
+                out-map="addressResponse"/>
+            <if condition="!addressResponse.addressMap"><return/></if>
```
Can we return with a meaningful message here?

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,145 @@
+
+            <!-- Searching State/Province Code -->
+            <entity-find-one entity-name="Geo" value-field="geoStateProvinceData">
+                <field-map field-name="geoId" from="postalAddress.stateProvinceGeoId"/>
```
Add check on geoTypeEnumId for state.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,145 @@
+
+            <!-- Searching Country Code -->
+            <entity-find-one entity-name="Geo" value-field="geoCountryData">
+                <field-map field-name="geoId" from="postalAddress.countryGeoId"/>
```
Add check on geoTypeEnumId for country.

ivanganatra committed:
Added return message on invalid addressMap,Added geoTypeEnumId condition on both state and country geoDetails,corrected contactMech id in return message when personName not found

ivanganatra committed:
Added x-locale parameterin inparameter of send#FedexRequest service

ivanganatra committed:
Added parameter map insted of x-locale while callin send#FedexRequest service

ivanganatra commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,156 @@
+            <set field="parameter" from="[xLocale:'en-US']" type="NewMap"/>
+            <service-call name="mantle.fedex.FedexServices.send#FedexRequest" out-map="response"
+                in-map="[shippingGatewayConfigId:shippingGatewayConfigId, method:'POST', path:'address/v1/addresses/resolve',
+                contentType:'application/json', requestMap:requestMap, parameter:parameter]"/>
```
parameter needs to be removed

ivanganatra commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,156 @@
+            <set field="requestMap" from="[addressesToValidate:[[address:addressResponse.addressMap.address]]]" type="NewMap"/>
+
+            <!-- Setting Connection With Fedex -->
+            <set field="parameter" from="[xLocale:'en-US']" type="NewMap"/>
```
parameter needs to be removed

ivanganatra committed:
Update FedexServices.xml

Removed x-locale

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,155 @@
+            <!-- Verifying Correctness of Address -->
+            <set field="responseMap" from="response.responseMap"/>
+            <set field="resolvedAddresses" from="responseMap?responseMap.output.resolvedAddresses[0]:null"/>
+            <set field="statusCodes" from="resolvedAddresses?resolvedAddresses.attributes:null"/>
```
Why the name statusCodes? I think it is not status codes which are getting returned here.

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,155 @@
+            <!-- Searching Country Code -->
+            <entity-find entity-name="Geo" list="geoCountryData">
+                <econdition field-name="geoId" from="postalAddress.countryGeoId"/>
+                <econditions combine="or">
```
I think we can skip the use of combine="or" as it is single condition here?

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,155 @@
+                <econdition field-name="geoId" from="postalAddress.stateProvinceGeoId"/>
+                <econditions combine="or">
+                    <econdition field-name="geoTypeEnumId" value="GEOT_STATE"/>
+                    <econdition field-name="geoTypeEnumId" value="GEOT_PROVINCE"/>
```
I think you can also use operator="in" as the condition is on the multiple values for the same field.

ivanganatra committed:
Change name, statusCodes to validationAttributes

ivanganatra committed:
Removed unused 'or' econdition from entity-find

ivanganatra committed:
Change operation from 'or' to 'in' in entity-find

gurveenbagga commented on service/mantle/fedex/FedexServices.xml:
```diff
@@ -0,0 +1,153 @@
+            <set field="responseMap" from="response.responseMap"/>
+            <set field="resolvedAddresses" from="responseMap?responseMap.output.resolvedAddresses[0]:null"/>
+            <set field="validationAttributes" from="resolvedAddresses?resolvedAddresses.attributes:null"/>
+            <if condition="responseMap==null || responseMap == '' || validationAttributes.InterpolatedStreetAddress=='true'
```
I think we should add null check for validationAttributes, if it is null, this could be potential NPE?

ivanganatra committed:
Added Null Pointer Check on ValidationAttributes variable
